### DISCUSSION TITLE
Duplicate Plugin Dependencies

### DISCUSSION THREAD

**Discussion by smessmer** [bevyengine/bevy#6418] [Q&A]
It seems `app.add_plugin(MyPlugin)` just directly calls `MyPlugin::build(app)`. This means, if the same plugin gets called multiple times from different places, its `build` function would be called multiple times. What are best practices around this? I imagine it could break things since bevy relies so heavily on global state (in the form of resources).

One probably common example I can think of is if you build a game that uses multiple plugins, say `PluginA` and `PluginB`, and each of those plugins rely on a GUI plugin, say bevy-egui. Who should call `app.add_plugin(bevy_egui::EguiPlugin)`? Should `PluginA` and `PluginB` not call that and instead document that they rely on it and that the main app needs to call `add_plugin(EguiPlugin)`? Should `PluginA` and `PluginB` both call `add_plugin(EguiPlugin)` and should `EguiPlugin` be written in a way that it can handle being called multiple times? Should bevy itself somehow handle this and deduplicate those calls?

**Comment by harudagondi** [comment]
There is a PR (#6411) that solves this issue by disallowing addition of plugins multiple times unless explicitly allowed.

IMO I think the user should add `EguiPlugin` and not `PluginA` or `PluginB` to be more explicit.

**Reply by mockersf** [reply]
Also, https://dev-docs.bevyengine.org/bevy/app/struct.App.html#method.is_plugin_added let you check if a plugin has already been added. `PluginA` and `PluginB` could check that `EguiPlugin` has not been added before adding it

**Reply by smessmer** [reply]
Oh interesting, the plugin registry seems to have been added after the currently newest release 0.8.1, hadn't seen it.

**Reply by mockersf** [reply]
yup it has been added very recently after the move away from resources as settings

**Comment by smessmer** [comment]
Taking a look at common dependencies, they seem to follow a pattern of "if plugin not added yet, add it myself":
```
        if !app.world.contains_resource::<EguiSettings>() {
            app.add_plugin(EguiPlugin);
        }
```
Sources:
[bevy-editor-pls](https://github.com/jakobhellermann/bevy_editor_pls/blob/63872f9d0ecbe94be1a3ebe4087797154f3bb00e/crates/bevy_editor_pls_core/src/editor.rs#L16)
[bevy-inspector-egui](https://github.com/jakobhellermann/bevy-inspector-egui/blob/39845201c31b7edbf41d4aaf5d324271b1cfa2ab/src/plugin.rs#L140)
[bevy-console](https://github.com/RichoDemus/bevy-console/blob/bf275328a2f4955e88b37f217d48af08e3519401/src/lib.rs#L43)

Is this the pattern recommended for libraries offering plugins that depend on other third party plugins?
